### PR TITLE
new getNodeForPopup method and more

### PR DESCRIPTION
**Pull request by daviddbal**
Tom,

This PR contains several commits that do the following:
- adds selectOneAppointmentCallback that is run when only one appointment is selected
- renames the callbacks I added to follow the verb-noun convention like the other callbacks in Agenda.
- adds -fx-fill colors to Agenda.css to allow me to use Rectangles as icons.
- adds getNodeForPopup and a map to back it in AgendaSkinTimeScale24HourAbstract.

The appointmentBodyPaneMap I added in the second commit has been removed.

-David

### CODE REVIEW

**Comment by tbee** [comment]
Where do you need the oneAppointmentCallback for? De appointments observable collection should provide the same info?

**Comment by daviddbal** [comment]
That's a good point.  I can listen to changes to selectedAppointments, if it contains only one then I can do what ever I want.  Is that what you mean?

**Comment by tbee** [comment]
yeah

**Comment by daviddbal** [comment]
I made the change.  The selecte one appointment callback is gone.

File: jfxtras-agenda/src/main/resources/jfxtras/internal/scene/control/skin/agenda/Agenda.css [hunk 1]
```diff
@@ -68,30 +68,30 @@
...
+.group0 { -fx-background-color: #AC725E; -fx-fill: #AC725E; }
+.group1 { -fx-background-color: #D06B64; -fx-fill: #D06B64; }
+.group2 { -fx-background-color: #F83A22; -fx-fill: #F83A22; }
+.group3 { -fx-background-color: #FA573C; -fx-fill: #FA573C; }
+.group4 { -fx-background-color: #FF7537; -fx-fill: #FF7537; }
+.group5 { -fx-background-color: #FFAD46; -fx-fill: #FFAD46; }
+.group6 { -fx-background-color: #42D692; -fx-fill: #42D692; }
+.group7 { -fx-background-color: #16A765; -fx-fill: #16A765; }
+.group8 { -fx-background-color: #7BD148; -fx-fill: #7BD148; }
+.group9 { -fx-background-color: #B3DC6C; -fx-fill: #B3DC6C; }
```
Comment (tbee) [inline jfxtras-agenda/src/main/resources/jfxtras/internal/scene/control/skin/agenda/Agenda.css:80]: Why are these fills needed?

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java [hunk 2]
```diff
@@ -238,7 +241,15 @@ public void refresh() {
...
 		nowUpdateRunnable.run(); 
-	}
+	}
+
+	/**
+	 * 
+	 */
+	@Override
+    public Pane getNodeForPopup(Appointment appointment) { return appointmentNodeMap.get(appointment); }
+    final private Map<Appointment, Pane> appointmentNodeMap = new WeakHashMap<>();
```
Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: I'm a bit worried about the hashcode method on the appointments and the hashmap being used here. What if one of the properties is changed after an appointment is put in this map?

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java [hunk 3]
```diff
@@ -252,7 +252,7 @@ private void setupDragging() {
 				handleDrag(appointment, dragPickupDateTime, dragDropDateTime);
 				
                 // has the client added a callback to process the change?
-                Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getAppointmentChangedCallback();
+                Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getChangedAppointmentCallback();
```
Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java]: This does not seem to belong in this PR. Please rebase and remove from the PR.

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/DurationDragger.java [hunk 4]
```diff
@@ -141,7 +141,7 @@ private void setupMouseDrag() {
 			appointmentPane.appointment.setEndLocalDateTime(endLocalDateTime);
 
             // has the client added a callback to process the change?
-            Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getAppointmentChangedCallback();
+            Callback<Appointment, Void> lChangedCallback = layoutHelp.skinnable.getChangedAppointmentCallback();
```
Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/DurationDragger.java]: This does not seem to belong in this PR. Please rebase and remove from the PR.

File: jfxtras-agenda/src/main/java/jfxtras/scene/control/agenda/Agenda.java [hunk 5]
```diff
@@ -349,17 +349,17 @@ private void constructAppointmentGroups() {
 	public void setEditAppointmentCallback(Callback<Appointment, Void> value) { this.editAppointmentCallbackObjectProperty.setValue(value); }
 	public Agenda withEditAppointmentCallback(Callback<Appointment, Void> value) { setEditAppointmentCallback(value); return this; }
 
-   /** appointmentChangedCallback:
+   /** changedAppointmentCallback:
```
Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/scene/control/agenda/Agenda.java]: This does not seem to belong in this PR. Please rebase and remove from the PR.

File: jfxtras-agenda/src/main/resources/jfxtras/internal/scene/control/skin/agenda/Agenda.css (hunk 1 above)
Comment (daviddbal) [inline jfxtras-agenda/src/main/resources/jfxtras/internal/scene/control/skin/agenda/Agenda.css:80]: I want to use rectangles as icons.  Rectangles allow some styling features, like rounded edges, that are not available to Panes.  Rectangles requires fills.

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java (hunk 2 above)
Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: Yes, that is a problem.  I can think of three choices:
1.  Not override equals and hashcode.
2.  Override equals only and not override hashcode.
3.  Not use a map for getNodeForPopup.

All options present problems.
1.  I need equals for unit testing.  I will have to use my own Appointment implementation for testing again.
2.  Unknown problems due to not following expected conventions.
3.  Other matching methods are less efficient than a map (e.g. scan through a list).

Can you think of other options?

Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: 1. Why do you need equals for unit testing? I've written agenda and a dozen other controls without it. You could write a static SomeUtil.isEqualTo(o1, o2). That said, Appointment is just an interface, so a user may provide his own implementation with equals en hashCode, so we cannot rely on these assumptions. The question would then be: why use a hashcode based map? 
   But then equals quickly becomes a problem. Maybe the best solution would be to use System.  identityHashCode(u) as the key, not a weak map and use a deletion listener on appointments to clean up the map.
2. not an option
3. I do not think a few simple loops will hurt that much in the grand scheme of things. But as said, you cannot use equals because there may two appointments with the same content, which we do not consider being the same.

We either decide to fully base ourselves on System.identityHashCode(u) or tell the user not to implement equals and hashcode.

Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: I am only using the equals method for unit testing.  All considered, it is too much trouble.  I will replace it with the static isEqualsTo method as you suggested.  That will eliminate my need for it.

I am not sure which is the better option for the getNodeForPopup method.  I need to study System.identityHashCode(u) first.

Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: identityHashCode always returns the hashCode of an object as it would if hashCode were not overridden, so two objects never are equal that way.

Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: I have read up on System.identityHashCode(u).  I think it should be used as the key as you suggested.  The deletion listener on appointments is easy enough to do.

You have a change listener in AllAppointments already.  Can I just add to that listener or do you think a whole new listener is best?

Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: Encapsulation would suggest a separate listener, since it would be part of a different class.

Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: I suppose it would be in a different class, AgendaSkinTimeScale24HourAbstract?  I'll have to think about it later.  It's too late for me and I need some sleep.  We are almost half-way around the world from each other.  You are just beginning your day.  I'm at the end of mine.

Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AgendaSkinTimeScale24HourAbstract.java]: I know. Isn't that great? And on my other chat is a New Sealander. 
Sleep well.

**Comment by daviddbal** [comment]
Tom,

I changed the appointmentNodeMap to use System.identityHashCode(a).  I added a clean-up listener also.  The feature passes simple manual testing.  I didn't add any automatic tests.

I also rebased.

Please take a look.

-David

**Comment by daviddbal** [comment]
Are the equals and hashcode methods staying or being removed?

**Comment by tbee** [comment]
Good question. On the one hand I'm a big fan of not having code that has no purpose, on the other hand, given that user may implement their own equals, it is good to have them in place to test against. But since Agenda must render two totally identical appointments separately, I think they should be removed.

**Comment by daviddbal** [comment]
Maybe they should be removed from Agenda, but put in a separate test-only Appointment implementation. That way some tests can be made to guarantee correct behavior in case a user implements another equals.

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java [hunk 6]
```diff
@@ -299,15 +299,17 @@ private void handleDrag(Agenda.Appointment appointment, LocalDateTime dragPickup
...
+                changed = true;
 			}
 			if (appointment.getEndLocalDateTime() != null) {
-				appointment.setEndLocalDateTime( appointment.getEndLocalDateTime().plus(duration) );
-				layoutHelp.callAppointmentChangedCallback(appointment);
-			}
+				appointment.setEndLocalDateTime( appointment.getEndLocalDateTime().plus(duration) );
+				changed = true;
+			}
+			if (changed) layoutHelp.callAppointmentChangedCallback(appointment);
```
Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:312]: A valid appointment can't start after it ends.  It also can't end before it starts.  Calling callAppointmentChangedCallback on changes to start and end separately can result in an invalid appointment after the first call, but before the second takes place.  This problem is avoided by updating both changes in one call.

My appointmenChangedCallback synchronizes the appointment to a VComponent.  It throws an exception when an invalid VComponent is created.

File: jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java [hunk 7]
```diff
@@ -333,18 +337,20 @@ else if ( (dragPickupInDayHeader && dragDropInDayBody) ) {
...
 			}
 			else {
 				// simply add the duration, but without time
 				Period period = Period.between(dragPickupDateTime.toLocalDate(), dragDropDateTime.toLocalDate());
 				appointment.setStartLocalDateTime( appointment.getStartLocalDateTime().toLocalDate().plus(period).atStartOfDay() );
 				appointment.setEndLocalDateTime( appointment.getEndLocalDateTime().toLocalDate().plus(period).plusDays(1).atStartOfDay() );
-				layoutHelp.callAppointmentChangedCallback(appointment);
+                changed = true;
 			}
+            if (changed) layoutHelp.callAppointmentChangedCallback(appointment);
```
Comment (tbee) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:353]: Recently I never use a single line if anymore, debugging is confusing then. Used to do it al lot though, so I'll fix this.

Comment (daviddbal) [inline jfxtras-agenda/src/main/java/jfxtras/internal/scene/control/skin/agenda/base24hour/AppointmentAbstractPane.java:353]: OK.  I'll make sure I don't use single line if in your code in the future.  I'll even think about adopting that convention too...
